### PR TITLE
[Metrics] class based embedding similarity + tests

### PR DESCRIPTION
**Pull request by SkafteNicki**
## What does this PR do?

PR https://github.com/PyTorchLightning/pytorch-lightning/pull/3349 added a new functional metric `embedding_similarity`. This PR just follows up and adds a class based interface + adds tests against sklearn

# Before submitting

- [ ] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests? 
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## PR review    
Anyone in the community is free to review the PR once the tests have passed.     
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: docs/source/metrics.rst [hunk 1]
```diff
@@ -158,6 +158,11 @@ DiceCoefficient
 .. autoclass:: pytorch_lightning.metrics.classification.DiceCoefficient
     :noindex:
 
+EmbeddingSimilarity
+^^^^^^^^^^^^^^^^^^^
+
+.. autoclass:: pytorch_lightning.metrics.self_supervised.EmbeddingSimilarity
+
```
Comment (awaelchli) [inline docs/source/metrics.rst:165]:
```suggestion
.. autoclass:: pytorch_lightning.metrics.self_supervised.EmbeddingSimilarity
    :noindex
    
```
this may solve the docs build error

File: docs/source/metrics.rst [hunk 2]
```diff
@@ -158,6 +158,12 @@ DiceCoefficient
 .. autoclass:: pytorch_lightning.metrics.classification.DiceCoefficient
     :noindex:
 
+EmbeddingSimilarity
+^^^^^^^^^^^^^^^^^^^
+
+.. autoclass:: pytorch_lightning.metrics.self_supervised.EmbeddingSimilarity
+    :noindex
```
Comment (awaelchli) [inline docs/source/metrics.rst:165]:
```suggestion
    :noindex:
```

File: pytorch_lightning/metrics/self_supervised.py [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+
+from pytorch_lightning.metrics.functional.self_supervised import embedding_similarity
+from pytorch_lightning.metrics.metric import TensorMetric
+
+
+class EmbeddingSimilarity(TensorMetric):
+    """
+
+
+    """
```
Comment (awaelchli) [inline pytorch_lightning/metrics/self_supervised.py:27]:
```suggestion
```

File: pytorch_lightning/metrics/self_supervised.py [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+                >>> embedding_similarity(embeddings)
+                tensor([[0.0000, 1.0000, 0.9759],
+                        [1.0000, 0.0000, 0.9759],
+                        [0.9759, 0.9759, 0.0000]])
+
+            Args:
+                similarity: 'dot' or 'cosine'
+                reduction: 'none', 'sum', 'mean' (all along dim -1)
+                zero_diagonal: if True, the diagonals are set to zero
+                reduce_group: the process group to reduce metric results from DDP
```
Comment (awaelchli) [inline pytorch_lightning/metrics/self_supervised.py:48]: there should be one less level of indentation to make it consistent with other docs, and otherwise the generated html has very strange formatting.

**Review by awaelchli — APPROVED** [review]
lgtm, just some whitespace that needs to be fixed for proper display in html

File: pytorch_lightning/metrics/self_supervised.py (hunk 3 above)
Comment (rohitgr7) [inline pytorch_lightning/metrics/self_supervised.py:27]:
```suggestion
    """
    Computes similarity between embeddings

    Example:
        >>> embeddings = torch.tensor([[1., 2., 3., 4.], [1., 2., 3., 4.], [4., 5., 6., 7.]])
        >>> embedding_similarity(embeddings)
        tensor([[0.0000, 1.0000, 0.9759],
                [1.0000, 0.0000, 0.9759],
                [0.9759, 0.9759, 0.0000]])
   
     """
```

File: pytorch_lightning/metrics/self_supervised.py [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+        """ Computes similarity between embeddings
+
+            Example:
+
+                >>> embeddings = torch.tensor([[1., 2., 3., 4.], [1., 2., 3., 4.], [4., 5., 6., 7.]])
+                >>> embedding_similarity(embeddings)
+                tensor([[0.0000, 1.0000, 0.9759],
+                        [1.0000, 0.0000, 0.9759],
+                        [0.9759, 0.9759, 0.0000]])
+
```
Comment (rohitgr7) [inline pytorch_lightning/metrics/self_supervised.py:43]:
```suggestion
    def __init__(
            self,
            similarity: str = 'cosine',
            zero_diagonal: bool = True,
            reduction: str = 'mean',
            reduce_group: Any = None
    ):
        """
```

File: pytorch_lightning/metrics/__init__.py [hunk 6]
```diff
@@ -56,4 +57,10 @@
     "SSIM"
 ]
 __sequence_metrics = ["BLEUScore"]
-__all__ = __regression_metrics + __classification_metrics + ["SklearnMetric"] + __sequence_metrics
+__selfsuper_metrics = ["EmbeddingSimilarity"]
+
+__all__ = __regression_metrics \
```
Comment (justusschock) [inline pytorch_lightning/metrics/__init__.py:62]: one question here: do we really have to define them here or can't we just import them from the respective file's __all__? (Something like `from pytorch_lightning.metrics.regression import __all__ as __regression_metrics`).

Also don't we have the __all__ defined in all the metrics files? If so we could just do from `pytorch_lightning.metrics.classification import *`

File: pytorch_lightning/metrics/self_supervised.py [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+# limitations under the License.
+
+from typing import Any, Optional, Sequence, Tuple
+
+import torch
+
+from pytorch_lightning.metrics.functional.self_supervised import embedding_similarity
+from pytorch_lightning.metrics.metric import TensorMetric
+
+
```
Comment (justusschock) [inline pytorch_lightning/metrics/self_supervised.py:22]:
```suggestion


__all__ = ['EmbeddingSimilarity']


```

File: pytorch_lightning/metrics/self_supervised.py (hunk 4 above)
Comment (SkafteNicki) [reply pytorch_lightning/metrics/self_supervised.py]: should be fixed now

File: pytorch_lightning/metrics/__init__.py (hunk 6 above)
Comment (SkafteNicki) [reply pytorch_lightning/metrics/__init__.py]: I have changed it to your suggestion, such that each file defines its own `__all__` which we then import in `__init__` and concatenate into one `__all__` for the hole package

**Review by justusschock — APPROVED** [review]

**Review by rohitgr7 — APPROVED** [review]

File: pytorch_lightning/metrics/self_supervised.py [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+        Args:
+            similarity: 'dot' or 'cosine'
+            reduction: 'none', 'sum', 'mean' (all along dim -1)
+            zero_diagonal: if True, the diagonals are set to zero
+            reduce_group: the process group to reduce metric results from DDP
+
+        """
+        super().__init__(name='embedding_similarity',
+                         reduce_group=reduce_group)
+        self.similarity = similarity
```
Comment (Borda) [inline pytorch_lightning/metrics/self_supervised.py:52]:
```suggestion
        assert similarity in ('dot', 'cosine')
        self.similarity = similarity
```

File: pytorch_lightning/metrics/self_supervised.py [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+        self.zero_diagonal = zero_diagonal
+        self.reduction = self.reduction
```
Comment (Borda) [inline pytorch_lightning/metrics/self_supervised.py:54]:
```suggestion
        asser reduction in ('none', 'sum', 'mean')
        self.reduction = reduction
```

File: pytorch_lightning/metrics/self_supervised.py [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+            similarity: 'dot' or 'cosine'
+            reduction: 'none', 'sum', 'mean' (all along dim -1)
+            zero_diagonal: if True, the diagonals are set to zero
+            reduce_group: the process group to reduce metric results from DDP
+
+        """
+        super().__init__(name='embedding_similarity',
+                         reduce_group=reduce_group)
+        self.similarity = similarity
+        self.zero_diagonal = zero_diagonal
```
Comment (Borda) [inline pytorch_lightning/metrics/self_supervised.py:53]:
```suggestion
        isinstance(zero_diagonal, bool)
        self.zero_diagonal = zero_diagonal
```

File: pytorch_lightning/metrics/self_supervised.py [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+            reduce_group: the process group to reduce metric results from DDP
+
+        """
+        super().__init__(name='embedding_similarity',
+                         reduce_group=reduce_group)
+        assert similarity in ('dot', 'cosine')
+        self.similarity = similarity
+        isinstance(zero_diagonal, bool)
+        self.zero_diagonal = zero_diagonal
+        asser reduction in ('none', 'sum', 'mean')
```
Comment (awaelchli) [inline pytorch_lightning/metrics/self_supervised.py:56]:
```suggestion
        assert reduction in ('none', 'sum', 'mean')
```

File: pytorch_lightning/metrics/metric.py [hunk 12]
```diff
@@ -287,3 +287,6 @@ def output_convert(self, data: Any, output: Any):
     def ddp_sync(self, data: Any, output: Any):
         return apply_to_collection(output, torch.Tensor, sync_ddp_if_available,
                                    self.reduce_group, self.reduce_op)
+
+
+__all__ = ["Metric", "TensorMetric", "NumpyMetric"]
```
[the remark this answers is not in the document]
Comment (awaelchli) [reply pytorch_lightning/metrics/metric.py:292]: codefactor also complains about this, not sure why. I think it should be fine though.

[the remark this answers is not in the document]
Comment (Borda) [reply pytorch_lightning/metrics/metric.py:292]: the point is that it is very related to importing from packages when you do not want to import all functions
https://stackoverflow.com/questions/44834/can-someone-explain-all-in-python

[the remark this answers is not in the document]
Comment (SkafteNicki) [reply pytorch_lightning/metrics/metric.py:292]: I agree that it is not a common place for `__all__`, normally I would put it at the top of the file, but then codefactor complains about it. But I can move it to the top if that is better.

[the remark this answers is not in the document]
Comment (Borda) [reply pytorch_lightning/metrics/metric.py:292]: yes, it is not very common to have it in other files than `__init__` so was there a reason to move it from the init?

[the remark this answers is not in the document]
Comment (SkafteNicki) [reply pytorch_lightning/metrics/metric.py:292]: I was changed due to a comment for justus at some point, but lets change it back since it is very uncommon practise

**Review by Borda — APPROVED** [review]
